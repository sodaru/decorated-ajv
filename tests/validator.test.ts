import {
  getAjv,
  getCompiledValidator,
  getValidator,
  validate
} from "../src/validator";
import Ajv from "ajv";
import { schema, schemaForCompile } from "./testSchema";
import { schema as complexSchema } from "./testComplexSchema";
import { join } from "path";
import { writeFile } from "fs/promises";

describe("Test JSON Validator", () => {
  test("getAjv", () => {
    const ajv = getAjv();
    expect(ajv).toBeInstanceOf(Ajv);
    expect(Object.keys(ajv.formats)).toEqual([
      "date",
      "time",
      "date-time",
      "duration",
      "uri",
      "uri-reference",
      "uri-template",
      "url",
      "email",
      "hostname",
      "ipv4",
      "ipv6",
      "regex",
      "uuid",
      "json-pointer",
      "json-pointer-uri-fragment",
      "relative-json-pointer",
      "byte",
      "int32",
      "int64",
      "float",
      "double",
      "password",
      "binary"
    ]);
  });

  test("getValidator", async () => {
    const validate = await getValidator({ type: "string" });
    expect(typeof validate).toEqual("function");
  });

  test("validate", async () => {
    await expect(
      validate({ type: "string" }, "this is awesome")
    ).resolves.toEqual([]);
  });

  test("validate for root error", async () => {
    await expect(validate(schema, null)).resolves.toMatchSnapshot();
  });

  test("validate for errorMessage at top", async () => {
    await expect(validate(schema, {})).resolves.toMatchSnapshot();
  });

  test("validate for errorMessage at 1 level down", async () => {
    await expect(
      validate(schema, { Type: "", Properties: {} })
    ).resolves.toMatchSnapshot();
  });

  test("validate for errorMessage at 2 level down", async () => {
    await expect(
      validate(schema, {
        Type: "SLP::DynamoDb::Item",
        DependsOn: {},
        Properties: { TableName: "Table1" }
      })
    ).resolves.toMatchSnapshot();
  });

  test("validate for errorMessage for additional Properties", async () => {
    await expect(
      validate(schema, {
        Type: "SLP::DynamoDb::Item",
        DependsOn: {
          resource: "my-resource"
        },
        Properties: {
          TableName: "Table1",
          "me@example.com": "This is my email"
        }
      })
    ).resolves.toMatchSnapshot();
  });

  test("validate with chain", async () => {
    await expect(
      validate(
        null,
        "this is awesome",
        await getValidator(schema, getAjv({ strictTuples: false }))
      )
    ).resolves.toMatchSnapshot();
  });
});

describe("Test getCompiledValidator", () => {
  test("getCompiledValidator", async () => {
    const compiled = await getCompiledValidator(schemaForCompile);
    expect(typeof compiled).toEqual("string");

    const compiledSchemaFilePath = join(__dirname, "compiledSchema.js");
    await writeFile(compiledSchemaFilePath, compiled);

    const compiledValidator = (await import(compiledSchemaFilePath)).default;

    expect(typeof compiledValidator).toEqual("function");

    expect(compiledValidator({})).toEqual(false);

    expect(compiledValidator.errors).toMatchSnapshot();

    expect(
      compiledValidator({
        attributes: { name: "Raghavendra", email: "this_must_fail" }
      })
    ).toEqual(false);

    expect(compiledValidator.errors).toMatchSnapshot();

    expect(
      compiledValidator({
        attributes: {
          name: "Raghavendra",
          email: "me@example.com",
          pic: "https://avatar.com/my-pic"
        }
      })
    ).toEqual(true);

    expect(compiledValidator.errors).toMatchSnapshot();
  });

  test("validate with compiledSchema", async () => {
    const compiledSchemaFilePath = join(__dirname, "compiledSchema.js");

    const compiledValidator = (await import(compiledSchemaFilePath)).default;

    await expect(validate(compiledValidator, {})).resolves.toMatchSnapshot();

    await expect(
      validate(compiledValidator, {
        attributes: { name: "Raghavendra", email: "this_must_fail" }
      })
    ).resolves.toMatchSnapshot();

    await expect(
      validate(compiledValidator, {
        attributes: {
          name: "Raghavendra",
          email: "me@example.com",
          pic: "https://avatar.com/my-pic"
        }
      })
    ).resolves.toEqual([]);
  });

  test("performance of compiled schema", async () => {
    const data = {
      attributes: {
        name: "Raghavendra",
        email: "me@example.com",
        pic: "https://avatar.com/my-pic"
      }
    };

    const compiledSchemaFilePath = join(__dirname, "compiledSchema.js");

    const compiledValidator = (await import(compiledSchemaFilePath)).default;

    const t1 = Date.now();
    compiledValidator(data);
    const t2 = Date.now();
    await validate(schemaForCompile, data);
    const t3 = Date.now();

    const timeWithCompiled = t2 - t1;
    const timeWithoutCompiled = t3 - t2;

    // eslint-disable-next-line no-console
    console.log({ timeWithCompiled, timeWithoutCompiled });

    expect(timeWithCompiled * 30).toBeLessThan(timeWithoutCompiled);
  });
});

describe("test for complex Schema", () => {
  test("match none of the oneOf options 1", async () => {
    await expect(validate(complexSchema, {})).resolves.toMatchSnapshot();
  });

  test("match none of the oneOf options 2", async () => {
    await expect(
      validate(complexSchema, { type: "aaaaa" })
    ).resolves.toMatchSnapshot();
  });

  test("partial match oneOf options 1", async () => {
    await expect(
      validate(complexSchema, { type: "initiateAuth" })
    ).resolves.toMatchSnapshot();
  });

  test("partial match oneOf options 2", async () => {
    await expect(
      validate(complexSchema, { type: "initiateAuth", challenge: {} })
    ).resolves.toMatchSnapshot();
  });

  test("internal match none of oneOf options", async () => {
    await expect(
      validate(complexSchema, {
        type: "respondToAuthChallenge",
        username: "fsadfd",
        challenge: {}
      })
    ).resolves.toMatchSnapshot();
  });

  test("internal partial match oneOf options", async () => {
    await expect(
      validate(complexSchema, {
        type: "respondToAuthChallenge",
        username: "fsadfd",
        challenge: {
          name: "PASSWORD_VERIFIER"
        }
      })
    ).resolves.toMatchSnapshot();
  });
});

describe("Test validator with async schema load", () => {
  test("test schema loaded from web", async () => {
    await expect(
      validate(
        {
          type: "object",
          properties: {
            inputs: {
              $ref: "https://form-input-schema.json-schema.sodaru.com/schemas/index.json#/properties/inputs"
            }
          }
        },
        { inputs: { url: { type: "text" } } }
      )
    ).resolves.toEqual([]);
  }, 20000);
});
