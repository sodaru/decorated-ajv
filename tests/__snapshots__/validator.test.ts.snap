// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test JSON Validator validate for errorMessage at 1 level down 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {},
    },
    "message": "must be SLP::DynamoDb::Item",
    "path": "Type",
  },
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "TableName",
      },
    },
    "message": "must have required property 'TableName'",
    "path": "Properties",
  },
]
`;

exports[`Test JSON Validator validate for errorMessage at 2 level down 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {},
    },
    "message": "must have resource",
    "path": "DependsOn",
  },
]
`;

exports[`Test JSON Validator validate for errorMessage at top 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {},
    },
    "message": "Must have Type and Properties",
    "path": "",
  },
]
`;

exports[`Test JSON Validator validate for errorMessage for additional Properties 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {},
    },
    "message": "Additional property \\"me@example.com\\" is not allowed",
    "path": "Properties.me@example.com",
  },
]
`;

exports[`Test JSON Validator validate for root error 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "type": "object",
      },
    },
    "message": "must be object",
    "path": "",
  },
]
`;

exports[`Test JSON Validator validate with chain 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "type": "object",
      },
    },
    "message": "must be object",
    "path": "",
  },
]
`;

exports[`Test getCompiledValidator getCompiledValidator 1`] = `
Array [
  Object {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'attributes'",
    "params": Object {
      "missingProperty": "attributes",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`Test getCompiledValidator getCompiledValidator 2`] = `
Array [
  Object {
    "instancePath": "/attributes/email",
    "keyword": "errorMessage",
    "message": "Enter a Valid Email Id",
    "params": Object {
      "errors": Array [
        Object {
          "emUsed": true,
          "instancePath": "/attributes/email",
          "keyword": "format",
          "message": "must match format \\"email\\"",
          "params": Object {
            "format": "email",
          },
          "schemaPath": "#/properties/attributes/allOf/1/properties/email/format",
        },
      ],
    },
    "schemaPath": "#/properties/attributes/allOf/1/properties/email/errorMessage",
  },
]
`;

exports[`Test getCompiledValidator getCompiledValidator 3`] = `null`;

exports[`Test getCompiledValidator validate with compiledSchema 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "attributes",
      },
    },
    "message": "must have required property 'attributes'",
    "path": "",
  },
]
`;

exports[`Test getCompiledValidator validate with compiledSchema 2`] = `
Array [
  Object {
    "context": Object {
      "params": Object {},
    },
    "message": "Enter a Valid Email Id",
    "path": "attributes.email",
  },
]
`;

exports[`test for complex Schema internal match none of oneOf options 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "name",
      },
    },
    "message": "must have required property 'name'",
    "path": "challenge",
  },
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "params",
      },
    },
    "message": "must have required property 'params'",
    "path": "challenge",
  },
]
`;

exports[`test for complex Schema internal partial match oneOf options 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "params",
      },
    },
    "message": "must have required property 'params'",
    "path": "challenge",
  },
]
`;

exports[`test for complex Schema match none of the oneOf options 1 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "type",
      },
    },
    "message": "must have required property 'type'",
    "path": "",
  },
]
`;

exports[`test for complex Schema match none of the oneOf options 2 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "allowedValues": Array [
          "initiateAuth",
          "respondToAuthChallenge",
          "signup",
          "confirmSignup",
          "resendConfirmationCode",
          "forgotPassword",
          "confirmForgotPassword",
        ],
      },
    },
    "message": "must be equal to one of the allowed values",
    "path": "type",
  },
]
`;

exports[`test for complex Schema partial match oneOf options 1 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "username",
      },
    },
    "message": "must have required property 'username'",
    "path": "",
  },
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "flow",
      },
    },
    "message": "must have required property 'flow'",
    "path": "",
  },
]
`;

exports[`test for complex Schema partial match oneOf options 2 1`] = `
Array [
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "username",
      },
    },
    "message": "must have required property 'username'",
    "path": "",
  },
  Object {
    "context": Object {
      "params": Object {
        "missingProperty": "flow",
      },
    },
    "message": "must have required property 'flow'",
    "path": "",
  },
  Object {
    "context": Object {
      "params": Object {
        "additionalProperty": "challenge",
      },
    },
    "message": "must NOT have additional properties",
    "path": "",
  },
]
`;
